File: tests/python/test_batch_operations.py
"""Unit tests for batch SDK methods (search_batch, count_batch, record, lookup, summary).

Tests use mocked HTTP responses to validate:
1. Correct URL construction for batch endpoints
2. Correct request payload structure
3. Proper response parsing via parse_* functions
4. Constraint validation (max 100 searches per batch)
5. Error handling for HTTP failures and invalid inputs
"""

import json
from unittest.mock import MagicMock, mock_open, patch

import pytest

from cli_generator import QueryBuilder


class TestBatchConstraints:
    """Test constraint validation for batch operations."""

    def test_search_batch_enforces_max_100_queries(self):
        """search_batch should reject >100 queries."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon") for _ in range(101)]

        with pytest.raises(ValueError, match="maximum 100 searches per batch request"):
            qb.search_batch(queries)

    def test_count_batch_enforces_max_100_queries(self):
        """count_batch should reject >100 queries."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon") for _ in range(101)]

        with pytest.raises(ValueError, match="maximum 100 searches per batch request"):
            qb.count_batch(queries)

    def test_search_batch_accepts_1_query(self):
        """search_batch should accept 1 query."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_resp = MagicMock()
            mock_resp.read.return_value = json.dumps({"status": {"success": True}, "results": [{"hits": 100}]}).encode(
                "utf-8"
            )
            mock_resp.__enter__.return_value = mock_resp
            mock_urlopen.return_value = mock_resp

            result = qb.search_batch(queries)
            assert isinstance(result, list)

    def test_search_batch_accepts_100_queries(self):
        """search_batch should accept exactly 100 queries."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon") for _ in range(100)]

        with patch("urllib.request.urlopen") as mock_urlopen:
            with patch("cli_generator.parse_batch_json") as mock_parse:
                mock_resp = MagicMock()
                results = [{"hits": i} for i in range(100)]
                response_data = {"status": {"success": True}, "results": results}
                mock_parse.return_value = json.dumps(response_data)
                mock_resp.read.return_value = json.dumps(response_data).encode("utf-8")
                mock_resp.__enter__.return_value = mock_resp
                mock_urlopen.return_value = mock_resp

                result = qb.search_batch(queries)
                assert len(result) == 100


class TestSearchBatchHTTPHandling:
    """Test HTTP request/response handling for search_batch."""

    def test_search_batch_constructs_correct_url(self):
        """search_batch should construct correct endpoint URL."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon").set_taxa(["Mammalia"])]

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_resp = MagicMock()
            mock_resp.read.return_value = json.dumps({"status": {"success": True}, "results": []}).encode("utf-8")
            mock_resp.__enter__.return_value = mock_resp
            mock_urlopen.return_value = mock_resp

            qb.search_batch(queries, api_base="http://localhost:3000/api")

            # Verify the URL was called
            call_args = mock_urlopen.call_args
            request_obj = call_args[0][0]
            assert "http://localhost:3000/api/v3/searchBatch" in request_obj.full_url

    def test_search_batch_uses_custom_api_version(self):
        """search_batch should respect custom api_version parameter."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_resp = MagicMock()
            mock_resp.read.return_value = json.dumps({"status": {"success": True}, "results": []}).encode("utf-8")
            mock_resp.__enter__.return_value = mock_resp
            mock_urlopen.return_value = mock_resp

            qb.search_batch(queries, api_base="http://localhost:3000/api", api_version="v4")

            call_args = mock_urlopen.call_args
            request_obj = call_args[0][0]
            assert "v4/searchBatch" in request_obj.full_url

    def test_search_batch_request_has_json_content_type(self):
        """search_batch should set Content-Type to application/json."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_resp = MagicMock()
            mock_resp.read.return_value = json.dumps({"status": {"success": True}, "results": []}).encode("utf-8")
            mock_resp.__enter__.return_value = mock_resp
            mock_urlopen.return_value = mock_resp

            qb.search_batch(queries)

            call_args = mock_urlopen.call_args
            request_obj = call_args[0][0]
            assert request_obj.headers.get("Content-type") == "application/json"

    def test_search_batch_payload_structure(self):
        """search_batch should send payload with 'searches' array."""
        qb = QueryBuilder("taxon")
        queries = [
            QueryBuilder("taxon").set_taxa(["Mammalia"]),
            QueryBuilder("taxon").set_taxa(["Aves"]),
        ]

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_resp = MagicMock()
            mock_resp.read.return_value = json.dumps({"status": {"success": True}, "results": [{}, {}]}).encode("utf-8")
            mock_resp.__enter__.return_value = mock_resp
            mock_urlopen.return_value = mock_resp

            qb.search_batch(queries)

            call_args = mock_urlopen.call_args
            request_obj = call_args[0][0]
            payload = json.loads(request_obj.data.decode("utf-8"))

            assert "searches" in payload
            assert len(payload["searches"]) == 2
            assert all("query_yaml" in s and "params_yaml" in s for s in payload["searches"])


class TestCountBatchHTTPHandling:
    """Test HTTP request/response handling for count_batch."""

    def test_count_batch_constructs_correct_url(self):
        """count_batch should construct correct endpoint URL."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon").set_taxa(["Mammalia"])]

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_resp = MagicMock()
            mock_resp.read.return_value = json.dumps(
                {"status": {"success": True, "hits": 100}, "results": [{"hits": 100}]}
            ).encode("utf-8")
            mock_resp.__enter__.return_value = mock_resp
            mock_urlopen.return_value = mock_resp

            qb.count_batch(queries)

            call_args = mock_urlopen.call_args
            request_obj = call_args[0][0]
            assert "v3/countBatch" in request_obj.full_url

    def test_count_batch_returns_hit_counts(self):
        """count_batch should return list of hit counts."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon"), QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            with patch("cli_generator.parse_batch_json") as mock_parse:
                mock_resp = MagicMock()
                response_data = {
                    "status": {"success": True},
                    "results": [
                        {"status": {"hits": 1000}},
                        {"status": {"hits": 2000}},
                    ],
                }
                mock_parse.return_value = json.dumps(response_data)
                mock_resp.read.return_value = json.dumps(response_data).encode("utf-8")
                mock_resp.__enter__.return_value = mock_resp
                mock_urlopen.return_value = mock_resp

                result = qb.count_batch(queries)

                assert result == [1000, 2000]


class TestRecordHTTPHandling:
    """Test HTTP request/response handling for record."""

    def test_record_constructs_correct_url(self):
        """record should construct correct endpoint URL."""
        qb = QueryBuilder("taxon").set_taxa(["9646"])

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_resp = MagicMock()
            mock_resp.read.return_value = json.dumps({"status": {"success": True}, "results": []}).encode("utf-8")
            mock_resp.__enter__.return_value = mock_resp
            mock_urlopen.return_value = mock_resp

            qb.record("9646", api_base="http://localhost:3000/api")

            call_args = mock_urlopen.call_args
            url_called = call_args[0][0]
            assert "v3/record" in url_called

    def test_record_uses_get_method(self):
        """record should use GET method with query params."""
        qb = QueryBuilder("taxon").set_taxa(["9646"])

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_resp = MagicMock()
            mock_resp.read.return_value = json.dumps({"status": {"success": True}, "results": []}).encode("utf-8")
            mock_resp.__enter__.return_value = mock_resp
            mock_urlopen.return_value = mock_resp

            qb.record("9646")

            call_args = mock_urlopen.call_args
            url_called = call_args[0][0]
            assert "recordId=9646" in url_called


class TestLookupHTTPHandling:
    """Test HTTP request/response handling for lookup."""

    def test_lookup_constructs_correct_url(self):
        """lookup should construct correct endpoint URL."""
        qb = QueryBuilder("taxon").set_taxa(["9646"])

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_resp = MagicMock()
            mock_resp.read.return_value = json.dumps({"status": {"success": True}, "results": []}).encode("utf-8")
            mock_resp.__enter__.return_value = mock_resp
            mock_urlopen.return_value = mock_resp

            qb.lookup("9646", api_base="http://localhost:3000/api")

            call_args = mock_urlopen.call_args
            url_called = call_args[0][0]
            assert "v3/lookup" in url_called


class TestSummaryHTTPHandling:
    """Test HTTP request/response handling for summary."""

    def test_summary_constructs_correct_url(self):
        """summary should construct correct endpoint URL."""
        qb = QueryBuilder("taxon").add_field("genome_size")

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_resp = MagicMock()
            mock_resp.read.return_value = json.dumps({"status": {"success": True}, "results": []}).encode("utf-8")
            mock_resp.__enter__.return_value = mock_resp
            mock_urlopen.return_value = mock_resp

            qb.summary("9646", "genome_size", api_base="http://localhost:3000/api")

            call_args = mock_urlopen.call_args
            url_called = call_args[0][0]
            assert "v3/summary" in url_called


class TestErrorHandling:
    """Test error handling in batch methods."""

    def test_search_batch_handles_http_error(self):
        """search_batch should propagate HTTP errors."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_urlopen.side_effect = Exception("HTTP 500: Server Error")

            with pytest.raises(Exception, match="HTTP 500"):
                qb.search_batch(queries)

    def test_count_batch_handles_http_error(self):
        """count_batch should propagate HTTP errors."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            mock_urlopen.side_effect = Exception("HTTP 500: Server Error")

            with pytest.raises(Exception, match="HTTP 500"):
                qb.count_batch(queries)

    def test_search_batch_handles_malformed_response(self):
        """search_batch should handle malformed JSON responses gracefully."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            with patch("cli_generator.parse_batch_json") as mock_parse:
                mock_resp = MagicMock()
                mock_resp.read.return_value = b"invalid json"
                mock_resp.__enter__.return_value = mock_resp
                mock_urlopen.return_value = mock_resp
                # parse_batch_json can transform the response; if it returns invalid JSON, json.loads will fail
                mock_parse.return_value = "not valid json"

                # This should raise an error since json.loads will fail
                with pytest.raises(json.JSONDecodeError):
                    qb.search_batch(queries)


class TestResponseParsing:
    """Test response parsing integration."""

    def test_search_batch_returns_results_array(self):
        """search_batch should return array of result objects."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon"), QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            with patch("cli_generator.parse_batch_json") as mock_parse:
                mock_resp = MagicMock()
                response_data = {
                    "status": {"success": True},
                    "results": [{"hits": 100, "result": "obj1"}, {"hits": 50, "result": "obj2"}],
                }
                mock_parse.return_value = json.dumps(response_data)
                mock_resp.read.return_value = json.dumps(response_data).encode("utf-8")
                mock_resp.__enter__.return_value = mock_resp
                mock_urlopen.return_value = mock_resp

                result = qb.search_batch(queries)

                assert isinstance(result, list)
                assert len(result) == 2
                assert result[0]["hits"] == 100
                assert result[1]["hits"] == 50

    def test_count_batch_extracts_hits_from_each_result(self):
        """count_batch should extract status.hits from each result."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon"), QueryBuilder("taxon"), QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            with patch("cli_generator.parse_batch_json") as mock_parse:
                mock_resp = MagicMock()
                response_data = {
                    "status": {"success": True},
                    "results": [
                        {"status": {"hits": 150}},
                        {"status": {"hits": 250}},
                        {"status": {"hits": 350}},
                    ],
                }
                mock_parse.return_value = json.dumps(response_data)
                mock_resp.read.return_value = json.dumps(response_data).encode("utf-8")
                mock_resp.__enter__.return_value = mock_resp
                mock_urlopen.return_value = mock_resp

                result = qb.count_batch(queries)

                assert result == [150, 250, 350]

    def test_search_batch_handles_empty_results(self):
        """search_batch should handle responses with no results."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            with patch("cli_generator.parse_batch_json") as mock_parse:
                mock_resp = MagicMock()
                response_data = {"status": {"success": True}, "results": []}
                mock_parse.return_value = json.dumps(response_data)
                mock_resp.read.return_value = json.dumps(response_data).encode("utf-8")
                mock_resp.__enter__.return_value = mock_resp
                mock_urlopen.return_value = mock_resp

                result = qb.search_batch(queries)

                assert result == []

    def test_count_batch_handles_empty_results(self):
        """count_batch should handle responses with no results."""
        qb = QueryBuilder("taxon")
        queries = [QueryBuilder("taxon")]

        with patch("urllib.request.urlopen") as mock_urlopen:
            with patch("cli_generator.parse_batch_json") as mock_parse:
                mock_resp = MagicMock()
                response_data = {"status": {"success": True}, "results": []}
                mock_parse.return_value = json.dumps(response_data)
                mock_resp.read.return_value = json.dumps(response_data).encode("utf-8")
                mock_resp.__enter__.return_value = mock_resp
                mock_urlopen.return_value = mock_resp

                result = qb.count_batch(queries)

                assert result == []

File: tests/javascript/test_batch_operations.mjs
/**
 * Unit tests for batch SDK methods (searchBatch, countBatch, record, lookup, summary).
 *
 * Tests validate via mocked fetch:
 * 1. Constraint enforcement (max 100 searches per batch)
 * 2. Correct URL construction for batch endpoints
 * 3. HTTP method (POST) and headers (Content-Type: application/json)
 * 4. Request payload structure (searches array with query_yaml, params_yaml)
 * 5. Error handling and response parsing
 *
 * Run via: node --test tests/javascript/test_batch_operations.mjs
 */

import { describe, it } from "node:test";

import assert from "node:assert/strict";

// Stub QueryBuilder for unit tests
class StubQueryBuilder {
  constructor(index) {
    this.index = index;
    this.taxa = [];
    this.fields = [];
  }

  set_taxa(taxa) {
    this.taxa = Array.isArray(taxa) ? taxa : [taxa];
    return this;
  }

  add_field(field) {
    this.fields.push(field);
    return this;
  }

  to_query_yaml() {
    const yaml = `index: ${this.index}\n`;
    if (this.taxa.length > 0) {
      return yaml + `taxa:\n${this.taxa.map((t) => `  - ${t}`).join("\n")}\n`;
    }
    return yaml;
  }

  to_params_yaml() {
    return `size: 5\n`;
  }

  async search_batch(queries, apiBase = "http://localhost:3000/api") {
    if (queries.length > 100) {
      throw new Error("maximum 100 searches per batch request");
    }
    const url = `${apiBase}/v3/search/batch`;
    const payload = {
      searches: queries.map((q) => ({
        query_yaml: q.to_query_yaml(),
        params_yaml: q.to_params_yaml(),
      })),
    };
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return JSON.parse(await response.text()).results || [];
  }

  async count_batch(queries, apiBase = "http://localhost:3000/api") {
    if (queries.length > 100) {
      throw new Error("maximum 100 searches per batch request");
    }
    const url = `${apiBase}/v3/count/batch`;
    const payload = {
      searches: queries.map((q) => ({
        query_yaml: q.to_query_yaml(),
        params_yaml: q.to_params_yaml(),
      })),
    };
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    const data = JSON.parse(await response.text());
    return (data.results || []).map((r) => r.status?.hits ?? 0);
  }

  async record(apiBase = "http://localhost:3000/api") {
    const url = `${apiBase}/v3/record`;
    const payload = {
      query_yaml: this.to_query_yaml(),
      params_yaml: this.to_params_yaml(),
    };
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return (JSON.parse(await response.text()).results || [])[0] || {};
  }

  async lookup(apiBase = "http://localhost:3000/api") {
    const url = `${apiBase}/v3/lookup`;
    const payload = {
      query_yaml: this.to_query_yaml(),
      params_yaml: this.to_params_yaml(),
    };
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return JSON.parse(await response.text()).results || {};
  }

  async summary(apiBase = "http://localhost:3000/api") {
    const url = `${apiBase}/v3/summary`;
    const payload = {
      query_yaml: this.to_query_yaml(),
      params_yaml: this.to_params_yaml(),
    };
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return JSON.parse(await response.text()).results || {};
  }
}

// ── Tests ─────────────────────────────────────────────────────────────────────

describe("Batch Operations - Constraints", async () => {
  it("search_batch should reject >100 queries", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = Array(101).fill(new StubQueryBuilder("taxon"));
    await assert.rejects(
      () => qb.search_batch(queries),
      /maximum 100 searches per batch request/,
    );
  });

  it("count_batch should reject >100 queries", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = Array(101).fill(new StubQueryBuilder("taxon"));
    await assert.rejects(
      () => qb.count_batch(queries),
      /maximum 100 searches per batch request/,
    );
  });

  it("search_batch should accept 1 query", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [new StubQueryBuilder("taxon")];
    const originalFetch = global.fetch;
    global.fetch = () =>
      Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({ status: { success: true }, results: [{}] }),
          ),
      });
    try {
      const result = await qb.search_batch(queries);
      assert.ok(Array.isArray(result));
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("search_batch should accept exactly 100 queries", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = Array(100).fill(new StubQueryBuilder("taxon"));
    const originalFetch = global.fetch;
    global.fetch = () =>
      Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({
              status: { success: true },
              results: Array(100).fill({}),
            }),
          ),
      });
    try {
      const result = await qb.search_batch(queries);
      assert.equal(result.length, 100);
    } finally {
      global.fetch = originalFetch;
    }
  });
});

describe("Batch Operations - Search Batch HTTP", async () => {
  it("search_batch should construct correct URL", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [new StubQueryBuilder("taxon").set_taxa(["Mammalia"])];
    let capturedUrl = null;
    const originalFetch = global.fetch;
    global.fetch = (url) => {
      capturedUrl = url;
      return Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({ status: { success: true }, results: [] }),
          ),
      });
    };
    try {
      await qb.search_batch(queries, "http://localhost:3000/api");
      assert.ok(
        capturedUrl.includes("http://localhost:3000/api/v3/search/batch"),
      );
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("search_batch should use POST method", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [qb];
    let capturedOptions = null;
    const originalFetch = global.fetch;
    global.fetch = (url, options) => {
      capturedOptions = options;
      return Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({ status: { success: true }, results: [] }),
          ),
      });
    };
    try {
      await qb.search_batch(queries);
      assert.equal(capturedOptions.method, "POST");
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("search_batch should set Content-Type header", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [qb];
    let capturedHeaders = null;
    const originalFetch = global.fetch;
    global.fetch = (url, options) => {
      capturedHeaders = options.headers;
      return Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({ status: { success: true }, results: [] }),
          ),
      });
    };
    try {
      await qb.search_batch(queries);
      assert.equal(capturedHeaders["Content-Type"], "application/json");
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("search_batch should send payload with searches array", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [
      new StubQueryBuilder("taxon").set_taxa(["Mammalia"]),
      new StubQueryBuilder("taxon").set_taxa(["Aves"]),
    ];
    let capturedPayload = null;
    const originalFetch = global.fetch;
    global.fetch = (url, options) => {
      capturedPayload = JSON.parse(options.body);
      return Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({ status: { success: true }, results: [{}, {}] }),
          ),
      });
    };
    try {
      await qb.search_batch(queries);
      assert.ok(capturedPayload.searches);
      assert.equal(capturedPayload.searches.length, 2);
      assert.ok("query_yaml" in capturedPayload.searches[0]);
      assert.ok("params_yaml" in capturedPayload.searches[0]);
    } finally {
      global.fetch = originalFetch;
    }
  });
});

describe("Batch Operations - Count Batch HTTP", async () => {
  it("count_batch should construct correct URL", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [new StubQueryBuilder("taxon").set_taxa(["Mammalia"])];
    let capturedUrl = null;
    const originalFetch = global.fetch;
    global.fetch = (url) => {
      capturedUrl = url;
      return Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({
              status: { success: true },
              results: [{ status: { hits: 100 } }],
            }),
          ),
      });
    };
    try {
      await qb.count_batch(queries, "http://localhost:3000/api");
      assert.ok(
        capturedUrl.includes("http://localhost:3000/api/v3/count/batch"),
      );
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("count_batch should return array of hit counts", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [
      new StubQueryBuilder("taxon"),
      new StubQueryBuilder("taxon"),
    ];
    const originalFetch = global.fetch;
    global.fetch = () =>
      Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({
              status: { success: true },
              results: [{ status: { hits: 1000 } }, { status: { hits: 2000 } }],
            }),
          ),
      });
    try {
      const result = await qb.count_batch(queries);
      assert.deepEqual(result, [1000, 2000]);
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("count_batch should handle missing hits gracefully", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [new StubQueryBuilder("taxon")];
    const originalFetch = global.fetch;
    global.fetch = () =>
      Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({ status: { success: true }, results: [{}] }),
          ),
      });
    try {
      const result = await qb.count_batch(queries);
      assert.deepEqual(result, [0]);
    } finally {
      global.fetch = originalFetch;
    }
  });
});

describe("Batch Operations - Record/Lookup/Summary HTTP", async () => {
  it("record should construct correct URL", async () => {
    const qb = new StubQueryBuilder("taxon").set_taxa(["9646"]);
    let capturedUrl = null;
    const originalFetch = global.fetch;
    global.fetch = (url) => {
      capturedUrl = url;
      return Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(JSON.stringify({ status: { success: true } })),
      });
    };
    try {
      await qb.record("http://localhost:3000/api");
      assert.ok(capturedUrl.includes("http://localhost:3000/api/v3/record"));
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("lookup should construct correct URL", async () => {
    const qb = new StubQueryBuilder("taxon").set_taxa(["9646"]);
    let capturedUrl = null;
    const originalFetch = global.fetch;
    global.fetch = (url) => {
      capturedUrl = url;
      return Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(JSON.stringify({ status: { success: true } })),
      });
    };
    try {
      await qb.lookup("http://localhost:3000/api");
      assert.ok(capturedUrl.includes("http://localhost:3000/api/v3/lookup"));
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("summary should construct correct URL", async () => {
    const qb = new StubQueryBuilder("taxon").add_field("genome_size");
    let capturedUrl = null;
    const originalFetch = global.fetch;
    global.fetch = (url) => {
      capturedUrl = url;
      return Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(JSON.stringify({ status: { success: true } })),
      });
    };
    try {
      await qb.summary("http://localhost:3000/api");
      assert.ok(capturedUrl.includes("http://localhost:3000/api/v3/summary"));
    } finally {
      global.fetch = originalFetch;
    }
  });
});

describe("Batch Operations - Error Handling", async () => {
  it("search_batch should throw on HTTP 500 error", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [new StubQueryBuilder("taxon")];
    const originalFetch = global.fetch;
    global.fetch = () =>
      Promise.resolve({
        ok: false,
        status: 500,
      });
    try {
      await assert.rejects(() => qb.search_batch(queries), /HTTP 500/);
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("count_batch should throw on HTTP 400 error", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [new StubQueryBuilder("taxon")];
    const originalFetch = global.fetch;
    global.fetch = () =>
      Promise.resolve({
        ok: false,
        status: 400,
      });
    try {
      await assert.rejects(() => qb.count_batch(queries), /HTTP 400/);
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("record should handle network errors", async () => {
    const qb = new StubQueryBuilder("taxon");
    const originalFetch = global.fetch;
    global.fetch = () => Promise.reject(new Error("Network connection failed"));
    try {
      await assert.rejects(() => qb.record(), /Network connection failed/);
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("search_batch should throw on malformed JSON response", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [new StubQueryBuilder("taxon")];
    const originalFetch = global.fetch;
    global.fetch = () =>
      Promise.resolve({
        ok: true,
        text: () => Promise.resolve("invalid json"),
      });
    try {
      await assert.rejects(() => qb.search_batch(queries));
    } finally {
      global.fetch = originalFetch;
    }
  });
});

describe("Batch Operations - Response Parsing", async () => {
  it("search_batch should return results list", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [
      new StubQueryBuilder("taxon"),
      new StubQueryBuilder("taxon"),
    ];
    const originalFetch = global.fetch;
    global.fetch = () =>
      Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({
              status: { success: true },
              results: [
                { hits: 100, data: "obj1" },
                { hits: 50, data: "obj2" },
              ],
            }),
          ),
      });
    try {
      const result = await qb.search_batch(queries);
      assert.ok(Array.isArray(result));
      assert.equal(result.length, 2);
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("search_batch should handle empty results", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [new StubQueryBuilder("taxon")];
    const originalFetch = global.fetch;
    global.fetch = () =>
      Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({ status: { success: true }, results: [] }),
          ),
      });
    try {
      const result = await qb.search_batch(queries);
      assert.equal(result.length, 0);
    } finally {
      global.fetch = originalFetch;
    }
  });

  it("count_batch should handle empty results", async () => {
    const qb = new StubQueryBuilder("taxon");
    const queries = [new StubQueryBuilder("taxon")];
    const originalFetch = global.fetch;
    global.fetch = () =>
      Promise.resolve({
        ok: true,
        text: () =>
          Promise.resolve(
            JSON.stringify({ status: { success: true }, results: [] }),
          ),
      });
    try {
      const result = await qb.count_batch(queries);
      assert.equal(result.length, 0);
    } finally {
      global.fetch = originalFetch;
    }
  });
});
